File: rust-test-node/src/env.rs
use std::env;
//use std::net::SocketAddr;
use std::sync::Arc;
use std::fs::OpenOptions;
use std::io::Write;
use tokio::time::{interval, Duration};
use prometheus_client::encoding::text::encode;

use warp::{Filter};
use tracing::{info, warn};
//use libp2p::metrics::Registry;

pub const MY_PORT: u16 = 5000;
pub const HTTP_PUBLISH_PORT: u16 = 8645;
pub const PROMETHEUS_PORT: u16 = 8008;

#[derive(Debug, Clone)]
pub struct PeerConfig {
    pub hostname: String,
    pub my_id: usize,
    pub network_size: usize,
    pub connect_to: usize,
    pub chunks: u32,
    pub muxer: String,
    pub address: String,
    pub service: String,
    pub in_shadow: bool,
}

pub fn get_peer_details() -> Result<PeerConfig, String> {

    let hostname = match gethostname::gethostname().into_string() {
        Ok(name) => name,
        Err(_) => return Err("Failed to get hostname".to_string()),
    };
    
    let my_id = hostname.split('-').nth(1)
            .and_then(|s| s.parse::<usize>().ok())
            .unwrap();
    
    let network_size = env::var("PEERS")
        .unwrap_or_else(|_| "100".to_string())
        .parse::<usize>()
        .unwrap_or(100);
    
    let connect_to = env::var("CONNECTTO")
        .unwrap_or_else(|_| "10".to_string())
        .parse::<usize>()
        .unwrap_or(10);
    
    let muxer = env::var("MUXER")
        .unwrap_or_else(|_| "yamux".to_string())
        .to_lowercase();

    let service = env::var("SERVICE")
        .unwrap_or_else(|_| "nimp2p-service".to_string());
    
    let in_shadow = env::var("SHADOWENV").is_ok();
    
    let address = match muxer.as_str() {
        "quic" => format!("/ip4/0.0.0.0/udp/{}/quic-v1", MY_PORT),
        _ => format!("/ip4/0.0.0.0/tcp/{}", MY_PORT),
    };
    
    let chunks = env::var("FRAGMENTS")
        .unwrap_or_else(|_| "1".to_string())
        .parse::<u32>()
        .unwrap_or(1);
    
    if !["quic", "yamux"].contains(&muxer.as_str()) {
        return Err(format!("Unknown muxer type: {}", muxer));
    }
    
    if connect_to >= network_size {
        return Err(format!("Not enough peers to make target connections. Network size: {}", network_size));
    }
    
    let config = PeerConfig {
        hostname, my_id, network_size, connect_to, chunks, muxer, address, service, in_shadow,
    };
    
    info!(
        "Host info hostname={}, peer={}, muxer={}, inShadow={}, address={}",
        config.hostname, config.my_id, config.muxer, config.in_shadow, config.address
    );
    
    Ok(config)
}

//Prometheus metrics
pub async fn start_metrics_server(registry: Arc<libp2p::metrics::Registry>) {

    let metrics = warp::path!("metrics")
        .and(warp::any().map(move || registry.clone()))
        .map(|registry: Arc<libp2p::metrics::Registry>| {

            let mut buffer = String::new();
            if let Err(e) = encode(&mut buffer, &registry) {
                warn!("Failed to encode metrics: {}", e);
                buffer = "Failed to encode metrics".to_string();
            }
            warp::reply::with_header(
                buffer,
                "Content-Type",
                "text/plain; version=0.0.4",
            )
        });
    
    let addr: std::net::SocketAddr = ([0, 0, 0, 0], PROMETHEUS_PORT).into();
    info!("Prometheus metrics server started at http://0.0.0.0:{}/metrics", PROMETHEUS_PORT);
    tokio::spawn(warp::serve(metrics).run(addr));
}

//log metrics if needed (useful for shadow simulations)
pub async fn store_metrics(
    registry_arc: Arc<libp2p::metrics::Registry>,
    peer_id: usize,
    interval_sec: u64,
) {

    tokio::time::sleep(Duration::from_millis((peer_id as u64) * 60)).await;
    let filename = format!("metrics_pod-{}.txt", peer_id);
    let mut interval = interval(Duration::from_secs(interval_sec));

    loop {
        interval.tick().await;

        let mut buffer = String::new();
        if let Err(e) = encode(&mut buffer, &registry_arc) {
            warn!("Failed to encode metrics for peer {}: {}", peer_id, e);
            continue;
        }

        match OpenOptions::new()
            .create(true)
            .append(true)
            .open(&filename)
        {
            Ok(mut file) => {
                writeln!(file, "\n# Metrics snapshot at {}", chrono::Utc::now()).ok();

                if let Err(e) = file.write_all(buffer.as_bytes()) {
                    warn!("Failed to write metrics for peer {}: {}", peer_id, e);
                } else {
                    info!("Metrics saved for peer {}", peer_id);
                }
            }
            Err(e) => {
                warn!("Failed to open metrics file for peer {}: {}", peer_id, e);
            }
        }
    }
}

File: nim-test-node/env.nim
import strutils, os, osproc
import chronos, metrics/chronos_httpserver, chronicles
from nativesockets import getHostname

let
  mountsMix* = existsEnv("MOUNTSMIX")                           #Full mix-net peer
  usesMix* = existsEnv("USESMIX")                               #Supports sending mix messages
  mixCount* = parseInt(getEnv("NUMMIX", "0"))                   #Number of mix peers (mountsMix + usesMix)
  inShadow* = getEnv("SHADOWENV").cmpIgnoreCase("true") == 0    #If Running for shadow simulator 
  httpPublishPort* = Port(8645)
  prometheusPort* = Port(8008)
  myPort* = Port(5000)
  chunks* = parseInt(getEnv("FRAGMENTS", "1"))                  #No. of fragments for each message
  mix_D* = parseInt(getEnv("MIXD", "4"))                        #No. of mix tunnels


proc getPeerDetails*(): Result[(int, int, int, string, string, string), string] =
  let 
    hostname = getHostname()
    myId = parseInt(hostname.split('-')[^1])
    networkSize = parseInt(getEnv("PEERS", "100"))
    connectTo = parseInt(getEnv("CONNECTTO", "10"))
    muxer = getEnv("MUXER", "yamux")
    filePath = if inShadow: "../" else: getEnv("FILEPATH", "./")
    address = if muxer.toLowerAscii() == "quic":
      "/ip4/0.0.0.0/udp/" & $myPort & "/quic-v1"
    else:
      "/ip4/0.0.0.0/tcp/" & $myPort
  
  if muxer.toLowerAscii() notin ["quic", "yamux", "mplex"]:
    return err("Unknown muxer type : " & muxer)

  if connectTo >= networkSize:
    return err("Not enough peers to make target connections. Network size : " & $networkSize)
  
  info "Host info ", hostname = hostname, peer = myId, muxer = muxer, mountsMix = mountsMix, usesMix = usesMix, mixCount = mixCount, inShadow = inShadow, address = address

  return ok((myId, networkSize, connectTo, muxer, filePath, address))

#Prometheus metrics
proc startMetricsServer*(
    serverIp: IpAddress, serverPort: Port
): Result[MetricsHttpServerRef, string] =
  info "Starting metrics HTTP server", serverIp = $serverIp, serverPort = $serverPort

  let metricsServerRes = MetricsHttpServerRef.new($serverIp, serverPort)
  if metricsServerRes.isErr():
    return err("metrics HTTP server start failed: " & $metricsServerRes.error)

  let server = metricsServerRes.value
  try:
    waitFor server.start()
  except CatchableError:
    return err("metrics HTTP server start failed: " & getCurrentExceptionMsg())

  info "Metrics HTTP server started", serverIp = $serverIp, serverPort = $serverPort
  ok(metricsServerRes.value)

#log metrics if needed (useful for shadow simulations)
proc storeMetrics*(myId: int) {.async.} =
  await sleepAsync((myId*60).milliseconds)
  while true:
    try:
      let cmd = "curl -s --connect-timeout 5 --max-time 5 http://localhost:" & 
          $prometheusPort & "/metrics >> metrics_pod-" & $myId & ".txt"
      
      let exitCode = execCmd(cmd)
      if exitCode == 0:
        info "Metrics saved for peer ", pod = myId
      else:
        info "Failed to fetch metrics for peer ", pod = myId, curlExitCode = $exitCode
    except CatchableError as e:
      info "Error storing metrics: ", error = e.msg
      return
    await sleepAsync(5.minutes)